### ISSUE TITLE
vpat 16-17: mouse actions alternative for virtualized table

### ISSUE THREAD

**Issue by abaevbog** [zotero/zotero#3970]
We want all functionality to be accessible via a keyboard and these things currently cannot be done in `itemTree` or `collectionTree` without using a mouse:
1. one can't set the primary sorting/resize/reorder/add/remove columns without right-clicking the header or dragging the rows
2. one can't move a collection into another collection or move a child note/attachment to be a child of another parent item without dragging

Similar problems are already solved with our context menus. For example, as an alternative to dragging an item to a collection, there is a "Add to collection" menu option. I think it would make sense to expand on this approach to make everything above available through the context menus.

In this case, the context menu of `itemTree` would need to be updated to allow one to not only add/remove but also resize/reorder/sort. Maybe each column in the context menu could become a submenu, so that one could, for example, do `contextMenu > creator > make wider/shorter/move right/move left`. This would be somewhat close to what happens with reordering of creators in `itemBox`.

In a similar fashion, we can add a "Move to collection" option to the context menu in `collectionTree` to allow collections to be moved into another collection/group without dragging. The question of avoiding drag-drop for to move a row as a child into another row in `itemTree` is a bit trickier, since we would need to list all available target items as `menuitem`s, and there theoretically could be thousands ... Maybe we could filter the items by some criteria (e.g. only use currently rendered items by virtualized table)?

One final problem to note is that there is no way to invoke the context menu to manipulate columns in `itemTree` without right-clicking on it, since the header is not focusable or selectable. Maybe we add a shortcut for this? As much as I personally dislike shortcuts, I can't think of any better way to expose this functionality without making the header selectable via keyboard which, I think, would make the interface of `itemTree` unnecessarily complex. Alternatively, we could add a small icon/button somewhere in the corner of the header that would be focusable but I'm not sure if it would look nice. 

These are some of my initial thoughts on how to address this. @dstillman @adomasven - what do you think?

**Comment by adomasven** [comment]
> one can't set the primary sorting/resize/reorder/add/remove columns without right-clicking the header or dragging the rows

Column sorting and visibility we can and should support via the View menu. Although we don't have a menubar in other views where the item tree is displayed (advanced search, citation dialog, select items). Not as big an issue on macOS with global menubars, but on Windows and Linux we might need to explore having a hiding menubar only visible on pressing Alt.

I haven't seen column resizing and reordering keyboard support for any app, partially because it's difficult to even implement, and presumably partially because for people with such vision impairment where using the mouse is not an option, column widths are not really relevant anyway, so at the very least we can split these particular actions out into a separate issue and think about whether, if and how to implement that.

> one can't move a collection into another collection

That's easy enough with a context menu option. Thunderbird also has a nice option "Move to [Folder] again", which we should probably support too both for collections and items.

> move a child note/attachment to be a child of another parent item without dragging

This seems untenable via a context menu, although if we did something like Spotify where they have a filter in the context menu, maybe? Seems like the most natural way to support this would be via cut/paste, and corresponding context menu options and keyboard shortcuts. The only problem is that it seems unnatural to support it without a copy action, even though you rarely if ever want to copy items in Zotero.

**Comment by dstillman** [comment]
> This seems untenable via a context menu, although if we did something like Spotify where they have a filter in the context menu, maybe?

I don't think we'd want to do it in the menu itself, but we could have a "Move to Item…" option that opened a simple panel that showed all non-trash items in the current library with pre-focused filter box at the top, and a "Move to Standalone Attachment" button in the bottom left that appeared for non-snapshot attachments. That would be relatively easy to build. Alternative would be a full library browser that showed only parent items, but that seems like a lot of effort for such a rare action. Or maybe we can just duplicate the Related Items window and it's trivial?

> Seems like the most natural way to support this would be via cut/paste, and corresponding context menu options and keyboard shortcuts. The only problem is that it seems unnatural to support it without a copy action, even though you rarely if ever want to copy items in Zotero.

I think the problem with this is that we care much more about using this mechanism for adding and moving items to collections, and using the same mechanism for two different things would be kind of confusing — you can't move a child item straight to another collection, so there's this complicated internal flowchart of whether you copied a parent or child item and whether a collection, parent item, or child item has focus when you paste. Better to just limit it to copying collections and items onto/into collections (which could behave the same even if the items list had focus). (It's also a little weird to paste onto a parent item to create a child item to begin with, since an item isn't really a container in the same way the collection is.)

(There's also the problem that Windows and macOS have different approaches for copying and moving. On Windows, it's Cut (move) or Copy (copy) and then Paste, while on macOS it's only Copy and then Paste or a modified Paste with Option to move. So even for just adding/moving to collections, we'd have to decide whether we were going to follow the platform convention.)

Agree with @adomasven on everything else.

**Comment by adomasven** [comment]
> I don't think we'd want to do it in the menu itself, but we could have a "Move to Item…" option that opened a simple panel that showed all non-trash items in the current library with pre-focused filter box at the top, and a "Move to Standalone Attachment" button in the bottom left that appeared for non-snapshot attachments. That would be relatively easy to build. Alternative would be a full library browser that showed only parent items, but that seems like a lot of effort for such a rare action. Or maybe we can just duplicate the Related Items window and it's trivial?

I didn't think about this, but yes, we have the generic select item dialog (used for Related Items, but not solely) that we could trivially adapt for this.

However the filter box could be useful for moving to collections still.

> (There's also the problem that Windows and macOS have different approaches for copying and moving. On Windows, it's Cut (move) or Copy (copy) and then Paste, while on macOS it's only Copy and then Paste or a modified Paste with Option to move. So even for just adding/moving to collections, we'd have to decide whether we were going to follow the platform convention.)

Right, and you also have to include context menu options for these too, complicating things still. I think just having a select items dialog when moving items and child items is good enough.

**Comment by abaevbog** [comment]
Cool! To briefly sum up my current direction here:
- View menu to house the options to manipulate the itemTree table, such as sort/add/remove columns. I agree that resizing of columns seems very awkward in this context, so I'll leave it for now and once all other pieces are in place, we can re-evaluate if it is still needed.
- a "Move to collection" context menu option to move a collection into another collection, similar to the existing option for items
- "Move to a parent item" (likely better named) context menu option for relevant items that will bring up the `selectItemsDialog.xhtml` dialog from where one can make the selection of the target parent. I like this, it feels like the most straightforward way to handle this.
- Explore the filter box for collections selection in both collectionTree and itemTree

**Comment by abaevbog** [comment]
These 3 linked PRs should expose all the functionality of itemTree and collectionTree discussed above to keyboard users (minus resizing of columns which can be tackled later if needed).

**Comment by tnajdek** [comment]
https://forums.zotero.org/discussion/122551/accessibility-copying-to-collection-in-different-library

**Comment by abaevbog** [comment]
Thank you, I was just looking at that post too! Copying an item to another library is addressed in https://github.com/zotero/zotero/pull/4785
